agent/local/state.go

```diff
@@ -0,0 +1,972 @@
+package local
+
+import (
+	"fmt"
+	"log"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/consul/agent/token"
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/consul/lib"
+	"github.com/hashicorp/consul/types"
+)
+
+// permissionDenied is returned when an ACL based rejection happens.
+const permissionDenied = "Permission denied"
+
+// Config is the configuration for the State. It is
+// populated during NewLocalAgent from the agent configuration to avoid
+// race conditions with the agent configuration.
+type Config struct {
+	AdvertiseAddr       string
+	CheckUpdateInterval time.Duration
+	Datacenter          string
+	NodeID              types.NodeID
+	NodeName            string
+	TaggedAddresses     map[string]string
+}
+
+// ServiceState describes the state of a service record.
+type ServiceState struct {
+	// Service is the local copy of the service record.
+	Service *structs.NodeService
+
+	// Token is the ACL to update the service record on the server.
```

Comment:
I'd say "update or delete" since that's kind of a subtle aspect of this.

Reply:
done